### PR TITLE
WIP [cpython] Conan v2 and version bump 3.10.4

### PR DESCRIPTION
**Pull request by jellespijker**
Specify library name and version:  **cpython/3.10.4**

Originally intended as a simple version bump, various linting errors and upgrades of dependencies forced this to grow in a full blown upgrade for Conan v2.

Still needs a lot of work 

Ultimaker Cura (https://github.com/Ultimaker/Cura) uses CPython 3.10.4 and we recently switched over to manage our complete build-process with Conan. This required us to bump up the version of this dependency and we want to contribute this change back to the community.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by jellespijker** [comment]
It seems that it can't find the module `conans` is this because that is an older module and expected to be deprecated for 2.0?

```
from conans import tools
ModuleNotFoundError: No module named 'conans'
```

Modernizing this recipe, just to have a minor version bump is a big scope creep. Unfortunately I don't have the time for this during my working hours, I might pick this up on on personal basis. But that would probably be during the weekend.

**Comment by jellespijker** [comment]
@ericLemanissier are the tests failing due to changes on my side? Or is modernizing to V2 and not using `conans.tools` a hard requirement?

```
Traceback (most recent call last):
  File "/home/conan/w/prod/BuildSingleReference/conan-center-index/recipes/cpython/all/test_package/setup.py", line 21, in <module>
    from conans import tools
ModuleNotFoundError: No module named 'conans'
```

**Comment by jellespijker** [comment]
Test keep failing for with the no `conans` module error. Should I create an issue for that failing CI?

**Comment by ericLemanissier** [comment]
I'm not sure why it fails, but it's a part which changed in this PR. You could stop importing conan.tools and use packaging.version.parse(version) instead: https://packaging.pypa.io/en/latest/version.html#packaging.version.parse

File: recipes/cpython/all/test_package/setup.py [hunk 1]
```diff
@@ -11,20 +11,23 @@
...
+use_2to3 = True
 if PY2:
     subdir = "py2"
     from distutils.core import setup, Extension
 elif PY3:
     subdir = "py3"
-    from setuptools import setup, Extension
+    from setuptools import setup, Extension, __version__ as setuptools_versions
+    from conans import tools
+    use_2to3 = tools.Version(setuptools_versions) <= tools.Version("58.0.0")
```
Comment (ericLemanissier) [inline recipes/cpython/all/test_package/setup.py:22]:
```suggestion
    import pkg_resources
    use_2to3 = pkg_resources.parse_version(setuptools_versions) <= pkg_resources.parse_version("58.0.0")
```

**Comment by jellespijker** [comment]
I see that this failed again in the clang linux build. Unfortunately, no time to work on this today, but I might boot of personal Linux laptop tomorrow.

**Comment by jellespijker** [comment]
See https://bugs.python.org/issue38443

It looks like 3.9.7 fails on Apple mac due the configure script not handling `--with-universal-archs= macOS` gracefully.
Which is strange because the fix for this has been backported to the 3.9 branch with a cherry-pick https://github.com/python/cpython/commit/748d38bf529d71d87cc7ef6d63e9df7d7d771ac9 

It does seem that this error isn't caused by changes introduce in this PR though. @ericLemanissier should it be solved in this?  It is hard for me to reproduce this on my Mac M1, Maybe one of my colleagues: @casperlamboo or @Joeydelarago can have a look tomorrow.

**Comment by ericLemanissier** [comment]
You are right that the failure is not the consequence of one of your change, cf https://github.com/conan-io/conan-center-index/pull/9891, however the bot won't merge the change if some configuration fails. If we understand why it fails and cannot solve it, we need to raise a ConanInvalidConfigurationException. There should be some explanations in `config.log` file in the build folder.

**Comment by jellespijker** [comment]
Is there a way for me to obtain that `config.log` since I can't reproduce this issue locally.

**Comment by ericLemanissier** [comment]
you could do something similar to
https://github.com/conan-io/conan-center-index/blob/master/recipes/qt/6.x.x/conanfile.py#L711-L720

**Comment by jellespijker** [comment]
That would actually be a nice feature to implement for more complex recipes like this one.

Might implement this today, if my calendar allows it, otherwise tomorrow

**Comment by jellespijker** [comment]
It also seems that I need to do some work to fix it for the v2 migration, looking at the linter. Will pick that up as well

**Comment by jellespijker** [comment]
Weird the previous build with Apple clang now succeed, while I technically didn't change anything..... But wait, maybe it failed the first time and the autool configure ran only once....

Anyhow I also get a new error in the 2.7 release

**Comment by ericLemanissier** [comment]
The error for both cpython versions is actually the same:
`clang: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=11.6'`. Some information is available on https://stackoverflow.com/questions/63972113/big-sur-clang-invalid-version-error-due-to-macosx-deployment-target. The easiest solution seems to be to define environment variable `SYSTEM_VERSION_COMPAT=1`

**Comment by jgsogo** [comment]
There are some errors reported by the linter, please, try to fix them (feel free to ask for help, if needeD). Thanks!

**Comment by jellespijker** [comment]
Help!

I didn't got around to it, I'm slowly walking up the dependency chain: m4, autoconf, automake, libffi, tcl, tk, ncursors, all of them needed to be updated first to get cpython work on Windows for us.

Maybe one of my colleagues @nallath or @casperlamboo could also take a look at those linter errors.

**Comment by uilianries** [comment]
@jellespijker please, consider https://github.com/Ultimaker/conan-center-index/pull/10 it should fix your linter errors

**Comment by uilianries** [comment]
@jellespijker More linter fixes, now related to the test_package: https://github.com/Ultimaker/conan-center-index/pull/11

File: recipes/cpython/all/conanfile.py [hunk 2]
```diff
@@ -317,15 +329,15 @@ def _upgrade_single_project_file(self, project_file):
...
         Restore it afterwards because it is needed to build some targets.
         """
-        tools.rename(os.path.join(self._source_subfolder, "PCbuild", "pcbuild.sln"),
+        rename(self, os.path.join(self._source_subfolder, "PCbuild", "pcbuild.sln"),
                      os.path.join(self._source_subfolder, "PCbuild", "pcbuild.sln.bak"))
-        tools.rename(os.path.join(self._source_subfolder, "PCbuild", "pcbuild.proj"),
+        rename(self, os.path.join(self._source_subfolder, "PCbuild", "pcbuild.proj"),
                      os.path.join(self._source_subfolder, "PCbuild", "pcbuild.proj.bak"))
         with tools.vcvars(self.settings):
             self.run("devenv \"{}\" /upgrade".format(project_file), run_environment=True)
```
Comment (ericLemanissier) [inline recipes/cpython/all/conanfile.py:337]:
```suggestion
            self.run(f"devenv \"{project_file}\" /upgrade", run_environment=True)
```

File: recipes/cpython/all/conanfile.py [hunk 3]
```diff
@@ -484,7 +496,7 @@ def _msvc_package_layout(self):
         python_args = " ".join("\"{}\"".format(a) for a in layout_args)
         self.run("{} {}".format(python_built, python_args), run_environment=True)
```
Comment (ericLemanissier) [inline recipes/cpython/all/conanfile.py:497]:
```suggestion
        python_args = " ".join(f"\"{a}\"" for a in layout_args)
        self.run(f"{python_built} {python_args}", run_environment=True)
```

File: recipes/cpython/all/test_package/conanfile.py [hunk 4]
```diff
@@ -113,7 +119,7 @@ def build(self):
                         "MSSdk": "1"
                     }
                     env.update(**AutoToolsBuildEnvironment(self).vars)
-                    with tools.environment_append(env):
+                    with legacy_tools.environment_append(env):
                         setup_args = [
                             "{}/setup.py".format(self.source_folder),
```
Comment (ericLemanissier) [inline recipes/cpython/all/test_package/conanfile.py:124]:
```suggestion
                            f"{self.source_folder}/setup.py",
```

File: recipes/cpython/all/conanfile.py [hunk 5]
```diff
@@ -616,7 +628,7 @@ def _lib_name(self):
         return "python{}{}".format(self._version_suffix, lib_ext)
 
     def _fix_install_name(self):
-        if tools.is_apple_os(self.settings.os) and self.options.shared:
+        if is_apple_os(self) and self.options.shared:
             buffer = StringIO()
             python = os.path.join(self.package_folder, "bin", "python")
             self.run('otool -L "%s"' % python, output=buffer)
```
Comment (ericLemanissier) [inline recipes/cpython/all/conanfile.py:634]:
```suggestion
            self.run(f'otool -L "{python}"', output=buffer)
```

**Comment by ericLemanissier** [comment]
just to be sure this message was not lost:

> The error for both cpython versions is actually the same: `clang: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=11.6'`. Some information is available on https://stackoverflow.com/questions/63972113/big-sur-clang-invalid-version-error-due-to-macosx-deployment-target. The easiest solution seems to be to define environment variable `SYSTEM_VERSION_COMPAT=1`

**Comment by jellespijker** [comment]
> just to be sure this message was not lost:
> 
> > The error for both cpython versions is actually the same: `clang: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=11.6'`. Some information is available on https://stackoverflow.com/questions/63972113/big-sur-clang-invalid-version-error-due-to-macosx-deployment-target. The easiest solution seems to be to define environment variable `SYSTEM_VERSION_COMPAT=1`

Thx I will take a look at this now

**Comment by ericLemanissier** [comment]
Does it work if you simply set environment variable SYSTEM_VERSION_COMPAT=1 ?

**Comment by jellespijker** [comment]
Gonna try that now

File: recipes/cpython/all/conanfile.py [hunk 6]
```diff
@@ -629,12 +647,14 @@ def _fix_install_name(self):
                     self.run("install_name_tool -change {} {} {}".format(library, new, python))
 
     def package_info(self):
+        self.cpp_info.includedirs = []
```
Comment (ericLemanissier) [inline recipes/cpython/all/conanfile.py:650]: This needs to be done first each component instead of the global one

Comment (jellespijker) [reply recipes/cpython/all/conanfile.py:650]: Yes I saw the the linter comment and figured as much.

I will add it to my Todo list for today

**Comment by ericLemanissier** [comment]
this bug looks like https://github.com/conan-io/conan/issues/11822, which should be solved in conan 1.52.0

**Comment by jellespijker** [comment]
> this bug looks like https://github.com/conan-io/conan/issues/11822, which should be solved in conan 1.52.0

Maybe because of the underscore in on of the component names?

**Comment by jellespijker** [comment]
I Think I'm gonna attempt to make the whole recipe Conan V2 ready, and cover my basis, making sure that the new generators are used consistently

**Comment by ericLemanissier** [comment]
conan has just been upgrade to 1.52.0 in https://github.com/conan-io/conan-center-index/commit/e93feccb608f0c3cae5d9d1e8cfa4b394161020c You can merge master into your branch to make CI use conan 1.52.0

**Pull request closed without merging**